release-notes/3.0/3.0.0-preview3.md

```diff
@@ -0,0 +1,176 @@
+}
+public class SqlCommand 
+{
+        public void RegisterColumnEncryptionKeyStoreProvidersOnCommand(IDictionary<string, SqlColumnEncryptionKeyStoreProvider> customProviders)
+}
+```
+
+The static API on `SqlConnection`, i.e. `SqlConnection.RegisterColumnEncryptionKeyStoreProviders` to register key store providers globally continues to be supported. The column encryption key cache maintained globally only applies to globally registered providers.
+
+#### Key Store Provider cache precedence
+
+System key store providers will be used from the global system cache if the requested provider is one. No providers should be registered on the connection or command instances if a system provider is needed.
+Registered key store providers are cached by the driver in 3 different cache layers. The global layer is as-is. The new per-connection and per-command level caches will be empty initially and can be set more than once.
+
+The precedence of the three caches will be as follows:
+
+- The per-command cache will be checked if it is not empty.
+- If the per-command cache is empty, the per-connection cache will be checked if it is not empty.
+- If the per-connection cache is empty, the global cache will be checked.
+
+Once a cache is found, the driver will **NOT** fall back to the other caches to search for requested provider.
+An exception will be thrown containing only the registered providers in the cache that was checked.
+
+#### Column Encryption Key cache precedence
+
+The column encryption keys (CEKs) for custom key store providers registered using the new instance-level APIs will not be cached by the driver. The key store providers shall implement their own cache to gain performance. This local cache of column encryption keys implemented by custom key store providers will be disabled by the driver if the key store provider instance is registered in the driver globally.
+
+A new API has also been introduced on the `SqlColumnEncryptionKeyStoreProvider` base class to set the cache time to live:
+
+```cs
+public abstract class SqlColumnEncryptionKeyStoreProvider
+{
+    // The default value of Column Encryption Key Cache Time to Live is 0.
+    // Provider's local cache is disabled for globally registered providers.
+    // Custom key store provider implementation must include column encryption key cache to provide caching support to locally registered providers.
+    public virtual TimeSpan? ColumnEncryptionKeyCacheTtl { get; set; } = new TimeSpan(0);
+}
+```
+
+### IP Address preference
+
+A new connection property `IPAddressPreference` is introduced to specify the IP address family preference to the driver when establishing TCP connection. Below are the three accepted values for this property:
+
+- **IPv4First**
+  - This is the default preference value. The driver will traverse IPv4 addresses first. If none of them can be connected successfully, it continues to try IPv6 addresses if there are any.
```

Comment:
```suggestion
  - This is the default preference value. The driver will use resolved IPv4 addresses first. If none of them can be connected to successfully, it will try resolved IPv6 addresses.
```